### DISCUSSION TITLE
Defer lookahead to allow optional tail

### DISCUSSION THREAD

**Discussion by ryo33** [jsinger67/parol#41] [Ideas]
I want to make a LL(1) language, and I've written the following grammar:

```
Expr: A B C
    | '!' Expr [ '->' Expr ]
    ;

A: "a";
B: "b";
C: "c";
```

`parol` emits this exceeds the maximum lookahead of 1.

Yes, it's not LL(1) because we should look ahead for `->` after the `Expr`, and the `Expr` can be the infinite size.
It's clear in an expanded grammar:

```
/* 0 */ Expr: A B C;
/* 1 */ Expr: '!' Expr ExprOpt /* Option */;
/* 2 */ ExprOpt /* Option<T>::Some */: '->' Expr;
/* 3 */ ExprOpt /* Option<T>::None */: ;
/* 4 */ A: "a";
/* 5 */ B: "b";
/* 6 */ C: "c";
```

But the production for `! abc -> abc` is the following struct:

```rust
pub struct BangExprArrowExpr<'t> {
    pub bang: Token<'t>, /* ! */
    pub expr: Box<Expr<'t>>,
    pub expr_opt: Option<Box<ExprOpt<'t>>>,
}
```

and not something like this:

```rust
pub enum BangExprArrowExpr {
    SomeTail(BangExprArrowExprSomeTail),
    NoneTail(BangExprArrowExprNoneTail),
}
```

Given this, it seems to me that LL(1) parsing can actually be done if we defer the lookahead of `->` because `!` already tells us this is a `BangExprArrowExpr`.

**Comment by jsinger67** [comment]
Quick shot, but please try this grammar:

```
Expr: A B C
    | '!' Expr '->' Expr
    | '!' Expr
    ;

A: "a";
B: "b";
C: "c";
```
This should make parol able to left factor it into a LL(1) grammar.

**Comment by ryo33** [comment]
It does indeed seem to output another expanded grammar, but it also exceeds the maximum lookahead of 5.
```
/* 0 */ Expr: '!' Expr ExprSuffix;
/* 1 */ Expr: A B C;
/* 2 */ ExprSuffix: '->' Expr;
/* 3 */ ExprSuffix: ;
/* 4 */ A: "a";
/* 5 */ B: "b";
/* 6 */ C: "c";
```

**Reply by ryo33** [reply]
Here is the source code: https://github.com/ryo33/parol_sandbox/tree/tail-optional

**Comment by jsinger67** [comment]
Ok, I will try it 😉

**Comment by jsinger67** [comment]
The problem is the recursion in `Expr`.

You have to formulate your grammar like this:

```parol
%start Expr
%title "ParolSandbox grammar"
%comment "Empty grammar generated by `parol`"
%scanner String {
    %auto_ws_off
    %auto_newline_off
 }

%%

Expr
    : AbcExpr
    | BangExpr
    ;

AbcExpr:
    | A B C;

BangExpr
    : '!' AbcExpr '->' Expr 
    | '!' AbcExpr
    ;

A: "a";
B: "b";
C: "c";
```

**Comment by jsinger67** [comment]
```
BangExpr
    : '!' AbcExpr ['->' Expr] 
    ;
```
will work this way too.

**Comment by ryo33** [comment]
```
Expr: A B C
    | '!' Expr [ '->' Other ]
    ;

A: "a";
B: "b";
C: "c";
Other: "other";
```

Somehow I think this example would work, but at least it's not an LL(1) grammar. Still, LL(1) parsing seems feasible, but I don't understand how PDA actually works, so I don't know if it is feasible. In the meantime, I'd like to start by taking a peek at the source code.

**Reply by ryo33** [reply]
I finally figured it out.
```
! ! ! abc -> other -> other
```

we cannot determine where the `->` belongs.

**Comment by jsinger67** [comment]
Yeah, you have basically a recursion at `Expr` that can infinitely add inner expressions.
You actually have to remove this recursion if your intended syntax allows this.
